src/dune_engine/source_tree.ml

```diff
@@ -511,6 +528,8 @@ end = struct
       | None, Some (path, _) -> Some path
     in
     Memo.Build.Option.map file ~f:(fun file ->
+        let open Memo.Build.O in
+        let* () = Memo.exec ensure_dune_project_file_exists project in
```

Comment:
`dune_file` is called for every directory, it would make more sense to do this check only at the root of the repo, which is the only directory for which Dune might invent a `dune-project` file. To do that, you could move the check to `Dune_project.load`.

Reply:
Doing that in `Dune_project.load` in the "infer" branch was my first attempt too, but it triggered warnings in too many situations. For example with this tree:

```
./
  a/
    dune-project
    dune
  b/
    dune-project
    dune
```

Running a command at the root would trigger a warning, because there's no dune-project there. But that is probably the right thing to do since there's no dune file to be interpreted either.

So I had a better look at what the original code was doing and found that it's done in `dune_file`. This PR aims to implement the the same logic:

https://github.com/ocaml/dune/blob/ce470d4aa6977f44969127dd90d2d492bbc68c62/src/dune_engine/file_tree.ml#L516-L518

I'll try if we can find a middle ground.

Reply:
I added a check that we're at the root, this removed the warning in a couple cases.

Reply:
With the new code, it seems that we don't trigger the warning in this case:

```
./
  a/
     dune
```

i.e. we only trigger a warning if there is a `dune` file at the root of the workspace. That seems a bit odd, wdyt?

Reply:
> I added a check that we're at the root, this removed the warning in a couple cases.

This check was not necessary. I removed it and added both your test case and Jeremie's.